sycl/test/extensions/annotated_usm/annotated_usm.cpp

```diff
@@ -0,0 +1,243 @@
+  auto APtr41 = malloc_device_annotated<decltype(P2)>(N, q);
+  auto APtr42 = malloc_device_annotated<decltype(P2), decltype(ANNP2)>(N, q);
+  static_assert(
+      std::is_same_v<decltype(APtr4), annotated_ptr<void, decltype(ANNP2)>>);
+  static_assert(std::is_same_v<decltype(APtr4), decltype(APtr41)>);
+  static_assert(std::is_same_v<decltype(APtr4), decltype(APtr42)>);
+
+  // APtr3 and APtr4 are not same because the underlying pointer type is int* vs
+  // void*
+  static_assert(!std::is_same_v<decltype(APtr4), decltype(APtr3)>);
+
+  // APtr1 and APtr2 do not have the same properties
+  static_assert(!std::is_same_v<decltype(APtr1), decltype(APtr2)>);
+
+  auto APtr5 = malloc_device_annotated<int>(N, q);
+  auto APtr51 = aligned_alloc_device_annotated<int>(128, N, q);
+  static_assert(
+      std::is_same_v<
+          decltype(APtr5),
+          annotated_ptr<int, decltype(properties{usm_kind<alloc::device>})>>);
+  static_assert(std::is_same_v<decltype(APtr5), decltype(APtr51)>);
+
+  auto APtr52 = malloc_device_annotated(N, q);
+  auto APtr53 = aligned_alloc_device_annotated(16, N, q);
+  static_assert(
+      std::is_same_v<
+          decltype(APtr52),
+          annotated_ptr<void, decltype(properties{usm_kind<alloc::device>})>>);
+  static_assert(std::is_same_v<decltype(APtr52), decltype(APtr53)>);
+
+  auto APtr6 = malloc_host_annotated<int>(N, q);
+  static_assert(
+      std::is_same_v<
+          decltype(APtr6),
+          annotated_ptr<int, decltype(properties{usm_kind<alloc::host>})>>);
+
+  auto APtr7 = malloc_shared_annotated<int>(N, q);
+  static_assert(
+      std::is_same_v<
+          decltype(APtr7),
+          annotated_ptr<int, decltype(properties{usm_kind<alloc::shared>})>>);
+
+  // The USM kinds differ
+  static_assert(!std::is_same_v<decltype(APtr6), decltype(APtr7)>);
+
+  free(APtr1, q);
+  free(APtr11, q);
+  free(APtr12, q);
+  free(APtr2, q);
+  free(APtr3, q);
+  free(APtr4, q);
+  free(APtr41, q);
+  free(APtr42, q);
+  free(APtr5, q);
+  free(APtr51, q);
+  free(APtr52, q);
+  free(APtr53, q);
+  free(APtr6, q);
+  free(APtr7, q);
+}
+
+void TestProperty(sycl::queue &q) {
+  // P4: input properties
+  properties P4{conduit, cache_config{large_slm}};
+  // ANNP4: output properties, with no usm_kind and no runtime property
+  properties ANNP4{conduit};
+  auto APtr1 = malloc_annotated<int>(N, q, alloc::device, P4);
+  static_assert(
+      std::is_same_v<decltype(APtr1), annotated_ptr<int, decltype(ANNP4)>>);
+
+  // APtr7 is of type annotated_ptr<int, decltype(properties{})>;
+  auto APtr2 = malloc_annotated<int>(N, q, alloc::device);
+  static_assert(std::is_same_v<decltype(APtr2), annotated_ptr<int>>);
+
+  properties P5{usm_kind<alloc::host>};
+  // Throws an exception with error code errc::invalid
+  // auto APtr8 = malloc_annotated<int>(N, q, sycl::usm::alloc::device, P5);
+
+  // Error: the USM kinds do not agree
+  // auto APtr9 = malloc_shared_annotated<int>(N, q, P5);
```

Comment:
Is it possible to automatically test these negative cases, and verify the errors or exceptions occur?